### PR TITLE
Add verify_batch_empty_fail_different_msg

### PR DESCRIPTION
**Pull request by jonas-lj**
Closing #97

### CODE REVIEW

**Comment by jonas-lj** [comment]
In this implementation, there's no error if the number of messages and pks/signatures does not match because it's using zip and the messages are represented by an iterator. Using zipeq will cause a panic if they do not match, but I'm not sure we want this either. Alternatively, the messages can be represented by a slice of slices, but I like the solution with an iterator better as it's more versatile.

File: fastcrypto/src/traits.rs [hunk 1]
```diff
@@ -110,6 +110,24 @@ pub trait VerifyingKey:
...
+    {
+        if sigs.is_empty() {
+            return Err(eyre!("Critical Error! This behaviour can signal something dangerous, and that someone may be trying to bypass signature verification through providing empty batches."));
+        }
+        if pks.len() != sigs.len() {
+            return Err(eyre!("Mismatch between number of signatures and public keys provided"));
+        }
+        pks.iter()
+            .zip(sigs)
+            .zip(msgs.into_iter())
```
Comment (jonas-lj) [inline fastcrypto/src/traits.rs:127]: Note that there's no error if the number of messages and signatures/keys do not match. Is this an issue?

File: fastcrypto/src/tests/bls12381_tests.rs [hunk 2]
```diff
@@ -102,6 +102,26 @@ fn signature_test_inputs() -> (Vec<u8>, Vec<BLS12381PublicKey>, Vec<BLS12381Sign
...
+        .zip(&digests)
+        .map(|(kp, digest)| {
+            let sig = kp.sign(digest.as_ref());
+            (kp.public().clone(), sig)
+        })
+        .unzip();
+
+    (digests, pubkeys, signatures)
+}
+
```
Comment (huitseeker) [inline fastcrypto/src/tests/bls12381_tests.rs:124]: Wouldn't this work?

```rust
fn signature_test_inputs_different_msg<S: Authenticator<PublicKey = PK>, PK: VerifyingKey<Sig = S>>
) -> (Vec<Vec<u8>>, Vec<PK>, Vec<S>) {
    // Make signatures.
    let digests: Vec<Vec<u8>> = [b"Hello", b"world", b"!!!!!"]
        .iter()
        .map(|msg| hash::Sha256::digest(*msg).to_vec())
        .collect();
    let (pubkeys, signatures): (Vec<_>, Vec<_>) = keys()
        .into_iter()
        .take(3)
        .zip(&digests)
        .map(|(kp, digest)| {
            let sig = kp.sign(digest.as_ref());
            (kp.public().clone(), sig)
        })
        .unzip();

    (digests, pubkeys, signatures)
}
```

**Review by huitseeker — COMMENTED** [review]
The tests look good, but perhaps the fixture can be shared between schemes by writing it generically?

File: fastcrypto/src/tests/bls12381_tests.rs (hunk 2 above)
Comment (jonas-lj) [reply fastcrypto/src/tests/bls12381_tests.rs:124]: That's a good idea. I think it could also work for many of the tests in BLS, ED25519 and secs256k1.

Comment (kchalkias) [reply fastcrypto/src/tests/bls12381_tests.rs:124]: Are we going to apply @huitseeker's suggestion in this PR?

Comment (jonas-lj) [reply fastcrypto/src/tests/bls12381_tests.rs:124]: Yea, I'll do that. Should we create a common_tests file for something like that?

Comment (kchalkias) [reply fastcrypto/src/tests/bls12381_tests.rs:124]: signature_tests for now

Comment (jonas-lj) [reply fastcrypto/src/tests/bls12381_tests.rs]: All the rebases made it hard to see what was actually added, so I've created a new PR, #119, and will close this one.

**Pull request closed without merging**